{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Making DBS input for SigProfiler"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The number of Shared SNVs between Passage B and A: 51,678\n",
      "The number of B specific SNVs: 6,577\n",
      "The number of A specific SNVs: 3,781\n",
      "The number of DBS variants for Passage B: 1,175\n",
      "The number of DBS variants for Passage B: 1,175\n"
     ]
    }
   ],
   "source": [
    "import io\n",
    "import os\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import gzip as gz\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib.ticker\n",
    "from matplotlib_venn import venn2, venn3\n",
    "\n",
    "sns.set_theme(font=\"Arial\", font_scale=1.15, style='ticks') \n",
    "matplotlib.rcParams['figure.dpi'] = 150\n",
    "plt.rc(\"axes.spines\", top=True, right=True)\n",
    "\n",
    "def read_vcf(path):\n",
    "    if path[-3:] == \".gz\": \n",
    "        with gz.open(path, 'rb') as f:\n",
    "            lines = [l.decode('utf-8') for l in f if not l.startswith(b'##')]\n",
    "            return pd.read_csv(\n",
    "                io.StringIO(''.join(lines)),\n",
    "                dtype={'#CHROM': str, 'POS': int, 'ID': str, 'REF': str, 'ALT': str,\n",
    "                       'QUAL': str, 'FILTER': str, 'INFO': str},\n",
    "                       sep='\\t'\n",
    "                       ).rename(columns={'#CHROM': 'CHROM'})\n",
    "    else:\n",
    "        with open(path, 'r') as f:\n",
    "            lines = [l for l in f if not l.startswith('##')]\n",
    "            return pd.read_csv(\n",
    "                io.StringIO(''.join(lines)),\n",
    "                dtype={'#CHROM': str, 'POS': int, 'ID': str, 'REF': str, 'ALT': str,\n",
    "                       'QUAL': str, 'FILTER': str, 'INFO': str},\n",
    "                       sep='\\t'\n",
    "                       ).rename(columns={'#CHROM': 'CHROM'})\n",
    "\n",
    "dir=\"/mmfs1/gscratch/stergachislab/mhsohny/SMaHT/Improving_SomaticVariantCalling_through_DSA/Fiber-seq\"\n",
    "DSA=\"/mmfs1/gscratch/stergachislab/mhsohny/SMaHT/DSA/DSA_COLO829BL_v3.0.0.fasta\"\n",
    "\n",
    "colotb_snvs_pass_annot_FlaggerHap_glfilt_pileupfilt = read_vcf(f\"{dir}/VariantCalls_DeepVariant_1.6.1/Mutational_Spectrum/COLO829T_PassageB_DSA.deepvariant.split.snv.modified.final.vcf.gz\")\n",
    "colota_snvs_pass_annot_FlaggerHap_glfilt_pileupfilt = read_vcf(f\"{dir}/VariantCalls_DeepVariant_1.6.1/Mutational_Spectrum/COLO829T_PassageA_DSA.deepvariant.split.snv.modified.final.vcf.gz\")\n",
    "\n",
    "colotb_snvs_pass_annot_FlaggerHap_glfilt_pileupfilt['SNVid'] = colotb_snvs_pass_annot_FlaggerHap_glfilt_pileupfilt[['CHROM', 'POS', 'REF', 'ALT']].astype(str).apply('_'.join, axis=1)\n",
    "colota_snvs_pass_annot_FlaggerHap_glfilt_pileupfilt['SNVid'] = colota_snvs_pass_annot_FlaggerHap_glfilt_pileupfilt[['CHROM', 'POS', 'REF', 'ALT']].astype(str).apply('_'.join, axis=1)\n",
    "\n",
    "colotb_snvs_pass_annot_FlaggerHap_glfilt_pileupfilt_set = set(colotb_snvs_pass_annot_FlaggerHap_glfilt_pileupfilt['SNVid'].values)\n",
    "colota_snvs_pass_annot_FlaggerHap_glfilt_pileupfilt_set = set(colota_snvs_pass_annot_FlaggerHap_glfilt_pileupfilt['SNVid'].values)\n",
    "\n",
    "tba_FlaggerHap_glfilt_pileupfilt_tba_set = colotb_snvs_pass_annot_FlaggerHap_glfilt_pileupfilt_set.intersection(colota_snvs_pass_annot_FlaggerHap_glfilt_pileupfilt_set)\n",
    "tb_FlaggerHap_glfilt_pileupfilt_onlytb_set = colotb_snvs_pass_annot_FlaggerHap_glfilt_pileupfilt_set.difference(colota_snvs_pass_annot_FlaggerHap_glfilt_pileupfilt_set)\n",
    "ta_FlaggerHap_glfilt_pileupfilt_onlyta_set = colota_snvs_pass_annot_FlaggerHap_glfilt_pileupfilt_set.difference(colotb_snvs_pass_annot_FlaggerHap_glfilt_pileupfilt_set)\n",
    "\n",
    "print(f\"The number of Shared SNVs between Passage B and A: {len(tba_FlaggerHap_glfilt_pileupfilt_tba_set):,}\")\n",
    "print(f\"The number of B specific SNVs: {len(tb_FlaggerHap_glfilt_pileupfilt_onlytb_set):,}\")\n",
    "print(f\"The number of A specific SNVs: {len(ta_FlaggerHap_glfilt_pileupfilt_onlyta_set):,}\")\n",
    "\n",
    "print(f\"The number of DBS variants for Passage B: {colotb_snvs_pass_annot_FlaggerHap_glfilt_pileupfilt[colotb_snvs_pass_annot_FlaggerHap_glfilt_pileupfilt.groupby(\"CHROM\")['POS'].diff() == 1].shape[0]:,}\")\n",
    "print(f\"The number of DBS variants for Passage B: {colotb_snvs_pass_annot_FlaggerHap_glfilt_pileupfilt[colotb_snvs_pass_annot_FlaggerHap_glfilt_pileupfilt.groupby(\"CHROM\")['POS'].diff().shift(-1) == 1].shape[0]:,}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>ID</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>QUAL</th>\n",
       "      <th>FILTER</th>\n",
       "      <th>INFO</th>\n",
       "      <th>FORMAT</th>\n",
       "      <th>COLO829T_PassageB_DSA</th>\n",
       "      <th>SNVid</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>117</th>\n",
       "      <td>haplotype1-0000001</td>\n",
       "      <td>12676857</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>42.4</td>\n",
       "      <td>PASS</td>\n",
       "      <td>.</td>\n",
       "      <td>GT:GQ:DP:AD:VAF:PL</td>\n",
       "      <td>0/1:12:117:23,94:0.803419:42,0,11</td>\n",
       "      <td>haplotype1-0000001_12676857_C_T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>220</th>\n",
       "      <td>haplotype1-0000001</td>\n",
       "      <td>28039723</td>\n",
       "      <td>.</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>63</td>\n",
       "      <td>PASS</td>\n",
       "      <td>.</td>\n",
       "      <td>GT:GQ:DP:AD:VAF:PL</td>\n",
       "      <td>1/1:56:112:0,112:1:62,57,0</td>\n",
       "      <td>haplotype1-0000001_28039723_A_G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>234</th>\n",
       "      <td>haplotype1-0000001</td>\n",
       "      <td>29598399</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>58.5</td>\n",
       "      <td>PASS</td>\n",
       "      <td>.</td>\n",
       "      <td>GT:GQ:DP:AD:VAF:PL</td>\n",
       "      <td>1/1:56:59:0,59:1:58,58,0</td>\n",
       "      <td>haplotype1-0000001_29598399_C_T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>260</th>\n",
       "      <td>haplotype1-0000001</td>\n",
       "      <td>35267980</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>72.7</td>\n",
       "      <td>PASS</td>\n",
       "      <td>.</td>\n",
       "      <td>GT:GQ:DP:AD:VAF:PL</td>\n",
       "      <td>1/1:62:93:0,93:1:72,62,0</td>\n",
       "      <td>haplotype1-0000001_35267980_G_A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>275</th>\n",
       "      <td>haplotype1-0000001</td>\n",
       "      <td>39873420</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>66.9</td>\n",
       "      <td>PASS</td>\n",
       "      <td>.</td>\n",
       "      <td>GT:GQ:DP:AD:VAF:PL</td>\n",
       "      <td>1/1:61:69:0,69:1:66,61,0</td>\n",
       "      <td>haplotype1-0000001_39873420_C_T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58070</th>\n",
       "      <td>haplotype2-0000079</td>\n",
       "      <td>152890010</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>61.7</td>\n",
       "      <td>PASS</td>\n",
       "      <td>.</td>\n",
       "      <td>GT:GQ:DP:AD:VAF:PL</td>\n",
       "      <td>1/1:60:64:0,64:1:61,66,0</td>\n",
       "      <td>haplotype2-0000079_152890010_C_T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58084</th>\n",
       "      <td>haplotype2-0000079</td>\n",
       "      <td>154873266</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>22.7</td>\n",
       "      <td>PASS</td>\n",
       "      <td>.</td>\n",
       "      <td>GT:GQ:DP:AD:VAF:PL</td>\n",
       "      <td>0/1:23:43:35,8:0.186047:22,0,46</td>\n",
       "      <td>haplotype2-0000079_154873266_G_A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58125</th>\n",
       "      <td>haplotype2-0000079</td>\n",
       "      <td>161251108</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>56.4</td>\n",
       "      <td>PASS</td>\n",
       "      <td>.</td>\n",
       "      <td>GT:GQ:DP:AD:VAF:PL</td>\n",
       "      <td>0/1:47:58:9,49:0.844828:56,0,47</td>\n",
       "      <td>haplotype2-0000079_161251108_C_T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58208</th>\n",
       "      <td>haplotype2-0000079</td>\n",
       "      <td>171322182</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>63</td>\n",
       "      <td>PASS</td>\n",
       "      <td>.</td>\n",
       "      <td>GT:GQ:DP:AD:VAF:PL</td>\n",
       "      <td>1/1:61:57:0,57:1:63,64,0</td>\n",
       "      <td>haplotype2-0000079_171322182_G_A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58228</th>\n",
       "      <td>haplotype2-0000136</td>\n",
       "      <td>14420</td>\n",
       "      <td>.</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>36.8</td>\n",
       "      <td>PASS</td>\n",
       "      <td>.</td>\n",
       "      <td>GT:GQ:DP:AD:VAF:PL</td>\n",
       "      <td>0/1:35:106:47,59:0.556604:36,0,39</td>\n",
       "      <td>haplotype2-0000136_14420_A_G</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1175 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    CHROM        POS ID REF ALT  QUAL FILTER INFO  \\\n",
       "117    haplotype1-0000001   12676857  .   C   T  42.4   PASS    .   \n",
       "220    haplotype1-0000001   28039723  .   A   G    63   PASS    .   \n",
       "234    haplotype1-0000001   29598399  .   C   T  58.5   PASS    .   \n",
       "260    haplotype1-0000001   35267980  .   G   A  72.7   PASS    .   \n",
       "275    haplotype1-0000001   39873420  .   C   T  66.9   PASS    .   \n",
       "...                   ...        ... ..  ..  ..   ...    ...  ...   \n",
       "58070  haplotype2-0000079  152890010  .   C   T  61.7   PASS    .   \n",
       "58084  haplotype2-0000079  154873266  .   G   A  22.7   PASS    .   \n",
       "58125  haplotype2-0000079  161251108  .   C   T  56.4   PASS    .   \n",
       "58208  haplotype2-0000079  171322182  .   G   A    63   PASS    .   \n",
       "58228  haplotype2-0000136      14420  .   A   G  36.8   PASS    .   \n",
       "\n",
       "                   FORMAT              COLO829T_PassageB_DSA  \\\n",
       "117    GT:GQ:DP:AD:VAF:PL  0/1:12:117:23,94:0.803419:42,0,11   \n",
       "220    GT:GQ:DP:AD:VAF:PL         1/1:56:112:0,112:1:62,57,0   \n",
       "234    GT:GQ:DP:AD:VAF:PL           1/1:56:59:0,59:1:58,58,0   \n",
       "260    GT:GQ:DP:AD:VAF:PL           1/1:62:93:0,93:1:72,62,0   \n",
       "275    GT:GQ:DP:AD:VAF:PL           1/1:61:69:0,69:1:66,61,0   \n",
       "...                   ...                                ...   \n",
       "58070  GT:GQ:DP:AD:VAF:PL           1/1:60:64:0,64:1:61,66,0   \n",
       "58084  GT:GQ:DP:AD:VAF:PL    0/1:23:43:35,8:0.186047:22,0,46   \n",
       "58125  GT:GQ:DP:AD:VAF:PL    0/1:47:58:9,49:0.844828:56,0,47   \n",
       "58208  GT:GQ:DP:AD:VAF:PL           1/1:61:57:0,57:1:63,64,0   \n",
       "58228  GT:GQ:DP:AD:VAF:PL  0/1:35:106:47,59:0.556604:36,0,39   \n",
       "\n",
       "                                  SNVid  \n",
       "117     haplotype1-0000001_12676857_C_T  \n",
       "220     haplotype1-0000001_28039723_A_G  \n",
       "234     haplotype1-0000001_29598399_C_T  \n",
       "260     haplotype1-0000001_35267980_G_A  \n",
       "275     haplotype1-0000001_39873420_C_T  \n",
       "...                                 ...  \n",
       "58070  haplotype2-0000079_152890010_C_T  \n",
       "58084  haplotype2-0000079_154873266_G_A  \n",
       "58125  haplotype2-0000079_161251108_C_T  \n",
       "58208  haplotype2-0000079_171322182_G_A  \n",
       "58228      haplotype2-0000136_14420_A_G  \n",
       "\n",
       "[1175 rows x 11 columns]"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "colotb_snvs_pass_annot_FlaggerHap_glfilt_pileupfilt[colotb_snvs_pass_annot_FlaggerHap_glfilt_pileupfilt.groupby(\"CHROM\")['POS'].diff() == 1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>ID</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>QUAL</th>\n",
       "      <th>FILTER</th>\n",
       "      <th>INFO</th>\n",
       "      <th>FORMAT</th>\n",
       "      <th>COLO829T_PassageB_DSA</th>\n",
       "      <th>SNVid</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>116</th>\n",
       "      <td>haplotype1-0000001</td>\n",
       "      <td>12676856</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>42.3</td>\n",
       "      <td>PASS</td>\n",
       "      <td>.</td>\n",
       "      <td>GT:GQ:DP:AD:VAF:PL</td>\n",
       "      <td>0/1:9:117:23,94:0.803419:41,0,8</td>\n",
       "      <td>haplotype1-0000001_12676856_C_T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>219</th>\n",
       "      <td>haplotype1-0000001</td>\n",
       "      <td>28039722</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>65.2</td>\n",
       "      <td>PASS</td>\n",
       "      <td>.</td>\n",
       "      <td>GT:GQ:DP:AD:VAF:PL</td>\n",
       "      <td>1/1:50:112:0,112:1:65,50,0</td>\n",
       "      <td>haplotype1-0000001_28039722_G_T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>233</th>\n",
       "      <td>haplotype1-0000001</td>\n",
       "      <td>29598398</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>63</td>\n",
       "      <td>PASS</td>\n",
       "      <td>.</td>\n",
       "      <td>GT:GQ:DP:AD:VAF:PL</td>\n",
       "      <td>1/1:56:59:0,59:1:63,57,0</td>\n",
       "      <td>haplotype1-0000001_29598398_C_T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>259</th>\n",
       "      <td>haplotype1-0000001</td>\n",
       "      <td>35267979</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>66.5</td>\n",
       "      <td>PASS</td>\n",
       "      <td>.</td>\n",
       "      <td>GT:GQ:DP:AD:VAF:PL</td>\n",
       "      <td>1/1:56:93:0,93:1:66,56,0</td>\n",
       "      <td>haplotype1-0000001_35267979_G_A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>274</th>\n",
       "      <td>haplotype1-0000001</td>\n",
       "      <td>39873419</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>64</td>\n",
       "      <td>PASS</td>\n",
       "      <td>.</td>\n",
       "      <td>GT:GQ:DP:AD:VAF:PL</td>\n",
       "      <td>1/1:58:69:0,69:1:64,59,0</td>\n",
       "      <td>haplotype1-0000001_39873419_C_T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58069</th>\n",
       "      <td>haplotype2-0000079</td>\n",
       "      <td>152890009</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>60.7</td>\n",
       "      <td>PASS</td>\n",
       "      <td>.</td>\n",
       "      <td>GT:GQ:DP:AD:VAF:PL</td>\n",
       "      <td>1/1:59:64:0,64:1:60,64,0</td>\n",
       "      <td>haplotype2-0000079_152890009_C_T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58083</th>\n",
       "      <td>haplotype2-0000079</td>\n",
       "      <td>154873265</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>18.5</td>\n",
       "      <td>PASS</td>\n",
       "      <td>.</td>\n",
       "      <td>GT:GQ:DP:AD:VAF:PL</td>\n",
       "      <td>0/1:19:43:35,8:0.186047:18,0,44</td>\n",
       "      <td>haplotype2-0000079_154873265_G_A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58124</th>\n",
       "      <td>haplotype2-0000079</td>\n",
       "      <td>161251107</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>53.8</td>\n",
       "      <td>PASS</td>\n",
       "      <td>.</td>\n",
       "      <td>GT:GQ:DP:AD:VAF:PL</td>\n",
       "      <td>0/1:42:58:9,49:0.844828:53,0,42</td>\n",
       "      <td>haplotype2-0000079_161251107_C_T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58207</th>\n",
       "      <td>haplotype2-0000079</td>\n",
       "      <td>171322181</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>60.7</td>\n",
       "      <td>PASS</td>\n",
       "      <td>.</td>\n",
       "      <td>GT:GQ:DP:AD:VAF:PL</td>\n",
       "      <td>1/1:55:57:0,57:1:60,56,0</td>\n",
       "      <td>haplotype2-0000079_171322181_G_A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58227</th>\n",
       "      <td>haplotype2-0000136</td>\n",
       "      <td>14419</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>36.6</td>\n",
       "      <td>PASS</td>\n",
       "      <td>.</td>\n",
       "      <td>GT:GQ:DP:AD:VAF:PL</td>\n",
       "      <td>0/1:30:106:47,59:0.556604:36,0,31</td>\n",
       "      <td>haplotype2-0000136_14419_G_C</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1175 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    CHROM        POS ID REF ALT  QUAL FILTER INFO  \\\n",
       "116    haplotype1-0000001   12676856  .   C   T  42.3   PASS    .   \n",
       "219    haplotype1-0000001   28039722  .   G   T  65.2   PASS    .   \n",
       "233    haplotype1-0000001   29598398  .   C   T    63   PASS    .   \n",
       "259    haplotype1-0000001   35267979  .   G   A  66.5   PASS    .   \n",
       "274    haplotype1-0000001   39873419  .   C   T    64   PASS    .   \n",
       "...                   ...        ... ..  ..  ..   ...    ...  ...   \n",
       "58069  haplotype2-0000079  152890009  .   C   T  60.7   PASS    .   \n",
       "58083  haplotype2-0000079  154873265  .   G   A  18.5   PASS    .   \n",
       "58124  haplotype2-0000079  161251107  .   C   T  53.8   PASS    .   \n",
       "58207  haplotype2-0000079  171322181  .   G   A  60.7   PASS    .   \n",
       "58227  haplotype2-0000136      14419  .   G   C  36.6   PASS    .   \n",
       "\n",
       "                   FORMAT              COLO829T_PassageB_DSA  \\\n",
       "116    GT:GQ:DP:AD:VAF:PL    0/1:9:117:23,94:0.803419:41,0,8   \n",
       "219    GT:GQ:DP:AD:VAF:PL         1/1:50:112:0,112:1:65,50,0   \n",
       "233    GT:GQ:DP:AD:VAF:PL           1/1:56:59:0,59:1:63,57,0   \n",
       "259    GT:GQ:DP:AD:VAF:PL           1/1:56:93:0,93:1:66,56,0   \n",
       "274    GT:GQ:DP:AD:VAF:PL           1/1:58:69:0,69:1:64,59,0   \n",
       "...                   ...                                ...   \n",
       "58069  GT:GQ:DP:AD:VAF:PL           1/1:59:64:0,64:1:60,64,0   \n",
       "58083  GT:GQ:DP:AD:VAF:PL    0/1:19:43:35,8:0.186047:18,0,44   \n",
       "58124  GT:GQ:DP:AD:VAF:PL    0/1:42:58:9,49:0.844828:53,0,42   \n",
       "58207  GT:GQ:DP:AD:VAF:PL           1/1:55:57:0,57:1:60,56,0   \n",
       "58227  GT:GQ:DP:AD:VAF:PL  0/1:30:106:47,59:0.556604:36,0,31   \n",
       "\n",
       "                                  SNVid  \n",
       "116     haplotype1-0000001_12676856_C_T  \n",
       "219     haplotype1-0000001_28039722_G_T  \n",
       "233     haplotype1-0000001_29598398_C_T  \n",
       "259     haplotype1-0000001_35267979_G_A  \n",
       "274     haplotype1-0000001_39873419_C_T  \n",
       "...                                 ...  \n",
       "58069  haplotype2-0000079_152890009_C_T  \n",
       "58083  haplotype2-0000079_154873265_G_A  \n",
       "58124  haplotype2-0000079_161251107_C_T  \n",
       "58207  haplotype2-0000079_171322181_G_A  \n",
       "58227      haplotype2-0000136_14419_G_C  \n",
       "\n",
       "[1175 rows x 11 columns]"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "colotb_snvs_pass_annot_FlaggerHap_glfilt_pileupfilt[colotb_snvs_pass_annot_FlaggerHap_glfilt_pileupfilt.groupby(\"CHROM\")['POS'].diff().shift(-1) == 1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
