{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Making DBS input for SigProfiler"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The number of Shared SNVs between Passage B and A: 51,678\n",
      "The number of B specific SNVs: 6,577\n",
      "The number of A specific SNVs: 3,781\n",
      "The number of DBS variants for Passage B: 1,175\n",
      "The number of DBS variants for Passage B: 1,175\n",
      "The number of DBS variants for Passage A: 1,097\n",
      "The number of DBS variants for Passage A: 1,097\n",
      "The number of DBS variants for Passage B: 1,175\n",
      "The number of DBS variants for Passage B: 1,175\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>ID</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>QUAL</th>\n",
       "      <th>FILTER</th>\n",
       "      <th>INFO</th>\n",
       "      <th>FORMAT</th>\n",
       "      <th>COLO829T_PassageB_DSA</th>\n",
       "      <th>SNVid</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>haplotype1-0000001</td>\n",
       "      <td>345669</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>65</td>\n",
       "      <td>PASS</td>\n",
       "      <td>.</td>\n",
       "      <td>GT:GQ:DP:AD:VAF:PL</td>\n",
       "      <td>1/1:59:133:2,131:0.984962:65,60,0</td>\n",
       "      <td>haplotype1-0000001_345669_C_T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>haplotype1-0000001</td>\n",
       "      <td>381602</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>70.4</td>\n",
       "      <td>PASS</td>\n",
       "      <td>.</td>\n",
       "      <td>GT:GQ:DP:AD:VAF:PL</td>\n",
       "      <td>1/1:65:131:0,131:1:70,66,0</td>\n",
       "      <td>haplotype1-0000001_381602_G_T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>haplotype1-0000001</td>\n",
       "      <td>436345</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>58.7</td>\n",
       "      <td>PASS</td>\n",
       "      <td>.</td>\n",
       "      <td>GT:GQ:DP:AD:VAF:PL</td>\n",
       "      <td>1/1:53:49:0,49:1:58,54,0</td>\n",
       "      <td>haplotype1-0000001_436345_G_A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>haplotype1-0000001</td>\n",
       "      <td>493800</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>68.7</td>\n",
       "      <td>PASS</td>\n",
       "      <td>.</td>\n",
       "      <td>GT:GQ:DP:AD:VAF:PL</td>\n",
       "      <td>1/1:63:83:0,83:1:68,63,0</td>\n",
       "      <td>haplotype1-0000001_493800_G_A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>haplotype1-0000001</td>\n",
       "      <td>547974</td>\n",
       "      <td>.</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>71.3</td>\n",
       "      <td>PASS</td>\n",
       "      <td>.</td>\n",
       "      <td>GT:GQ:DP:AD:VAF:PL</td>\n",
       "      <td>1/1:65:123:0,123:1:71,66,0</td>\n",
       "      <td>haplotype1-0000001_547974_A_G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58250</th>\n",
       "      <td>haplotype2-0000137</td>\n",
       "      <td>581230</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>5.5</td>\n",
       "      <td>PASS</td>\n",
       "      <td>.</td>\n",
       "      <td>GT:GQ:DP:AD:VAF:PL</td>\n",
       "      <td>0/1:6:127:93,27:0.212598:4,0,37</td>\n",
       "      <td>haplotype2-0000137_581230_C_T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58251</th>\n",
       "      <td>haplotype2-0000137</td>\n",
       "      <td>606494</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>31.3</td>\n",
       "      <td>PASS</td>\n",
       "      <td>.</td>\n",
       "      <td>GT:GQ:DP:AD:VAF:PL</td>\n",
       "      <td>0/1:13:250:162,88:0.352:31,0,12</td>\n",
       "      <td>haplotype2-0000137_606494_G_A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58252</th>\n",
       "      <td>haplotype2-0000194</td>\n",
       "      <td>22585</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>44</td>\n",
       "      <td>PASS</td>\n",
       "      <td>.</td>\n",
       "      <td>GT:GQ:DP:AD:VAF:PL</td>\n",
       "      <td>0/1:40:61:14,46:0.754098:43,0,42</td>\n",
       "      <td>haplotype2-0000194_22585_G_A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58253</th>\n",
       "      <td>haplotype2-0000194</td>\n",
       "      <td>36807</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>45.9</td>\n",
       "      <td>PASS</td>\n",
       "      <td>.</td>\n",
       "      <td>GT:GQ:DP:AD:VAF:PL</td>\n",
       "      <td>0/1:23:60:11,49:0.816667:45,0,23</td>\n",
       "      <td>haplotype2-0000194_36807_C_T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58254</th>\n",
       "      <td>haplotype2-0000194</td>\n",
       "      <td>37780</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>40</td>\n",
       "      <td>PASS</td>\n",
       "      <td>.</td>\n",
       "      <td>GT:GQ:DP:AD:VAF:PL</td>\n",
       "      <td>0/1:10:57:8,49:0.859649:39,0,9</td>\n",
       "      <td>haplotype2-0000194_37780_C_T</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>58255 rows Ã— 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    CHROM     POS ID REF ALT  QUAL FILTER INFO  \\\n",
       "0      haplotype1-0000001  345669  .   C   T    65   PASS    .   \n",
       "1      haplotype1-0000001  381602  .   G   T  70.4   PASS    .   \n",
       "2      haplotype1-0000001  436345  .   G   A  58.7   PASS    .   \n",
       "3      haplotype1-0000001  493800  .   G   A  68.7   PASS    .   \n",
       "4      haplotype1-0000001  547974  .   A   G  71.3   PASS    .   \n",
       "...                   ...     ... ..  ..  ..   ...    ...  ...   \n",
       "58250  haplotype2-0000137  581230  .   C   T   5.5   PASS    .   \n",
       "58251  haplotype2-0000137  606494  .   G   A  31.3   PASS    .   \n",
       "58252  haplotype2-0000194   22585  .   G   A    44   PASS    .   \n",
       "58253  haplotype2-0000194   36807  .   C   T  45.9   PASS    .   \n",
       "58254  haplotype2-0000194   37780  .   C   T    40   PASS    .   \n",
       "\n",
       "                   FORMAT              COLO829T_PassageB_DSA  \\\n",
       "0      GT:GQ:DP:AD:VAF:PL  1/1:59:133:2,131:0.984962:65,60,0   \n",
       "1      GT:GQ:DP:AD:VAF:PL         1/1:65:131:0,131:1:70,66,0   \n",
       "2      GT:GQ:DP:AD:VAF:PL           1/1:53:49:0,49:1:58,54,0   \n",
       "3      GT:GQ:DP:AD:VAF:PL           1/1:63:83:0,83:1:68,63,0   \n",
       "4      GT:GQ:DP:AD:VAF:PL         1/1:65:123:0,123:1:71,66,0   \n",
       "...                   ...                                ...   \n",
       "58250  GT:GQ:DP:AD:VAF:PL    0/1:6:127:93,27:0.212598:4,0,37   \n",
       "58251  GT:GQ:DP:AD:VAF:PL    0/1:13:250:162,88:0.352:31,0,12   \n",
       "58252  GT:GQ:DP:AD:VAF:PL   0/1:40:61:14,46:0.754098:43,0,42   \n",
       "58253  GT:GQ:DP:AD:VAF:PL   0/1:23:60:11,49:0.816667:45,0,23   \n",
       "58254  GT:GQ:DP:AD:VAF:PL     0/1:10:57:8,49:0.859649:39,0,9   \n",
       "\n",
       "                               SNVid  \n",
       "0      haplotype1-0000001_345669_C_T  \n",
       "1      haplotype1-0000001_381602_G_T  \n",
       "2      haplotype1-0000001_436345_G_A  \n",
       "3      haplotype1-0000001_493800_G_A  \n",
       "4      haplotype1-0000001_547974_A_G  \n",
       "...                              ...  \n",
       "58250  haplotype2-0000137_581230_C_T  \n",
       "58251  haplotype2-0000137_606494_G_A  \n",
       "58252   haplotype2-0000194_22585_G_A  \n",
       "58253   haplotype2-0000194_36807_C_T  \n",
       "58254   haplotype2-0000194_37780_C_T  \n",
       "\n",
       "[58255 rows x 11 columns]"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import io\n",
    "import os\n",
    "from tqdm import tqdm\n",
    "import pysam\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import gzip as gz\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib.ticker\n",
    "from matplotlib_venn import venn2, venn3\n",
    "import sigProfilerPlotting as sigPlt\n",
    "from SigProfilerAssignment import Analyzer as Analyze\n",
    "\n",
    "sns.set_theme(font=\"Arial\", font_scale=1.15, style='ticks') \n",
    "matplotlib.rcParams['figure.dpi'] = 150\n",
    "plt.rc(\"axes.spines\", top=True, right=True)\n",
    "\n",
    "def read_vcf(path):\n",
    "    if path[-3:] == \".gz\": \n",
    "        with gz.open(path, 'rb') as f:\n",
    "            lines = [l.decode('utf-8') for l in f if not l.startswith(b'##')]\n",
    "            return pd.read_csv(\n",
    "                io.StringIO(''.join(lines)),\n",
    "                dtype={'#CHROM': str, 'POS': int, 'ID': str, 'REF': str, 'ALT': str,\n",
    "                       'QUAL': str, 'FILTER': str, 'INFO': str},\n",
    "                       sep='\\t'\n",
    "                       ).rename(columns={'#CHROM': 'CHROM'})\n",
    "    else:\n",
    "        with open(path, 'r') as f:\n",
    "            lines = [l for l in f if not l.startswith('##')]\n",
    "            return pd.read_csv(\n",
    "                io.StringIO(''.join(lines)),\n",
    "                dtype={'#CHROM': str, 'POS': int, 'ID': str, 'REF': str, 'ALT': str,\n",
    "                       'QUAL': str, 'FILTER': str, 'INFO': str},\n",
    "                       sep='\\t'\n",
    "                       ).rename(columns={'#CHROM': 'CHROM'})\n",
    "\n",
    "def reverse_complement(string):\n",
    "    try:\n",
    "        complement_dict = {'A':'T', 'T':'A', 'G':'C', 'C':'G'}\n",
    "        complement_string = ''.join([complement_dict[s] for s in string])\n",
    "    except KeyError:\n",
    "        raise ValueError(\"Invalid character other than A,T,G and C\")\n",
    "    return complement_string[::-1]\n",
    "\n",
    "def dbs_context_change(string):\n",
    "    # GC>CT => GC>AG\n",
    "    # 5'-GC-3' > 5'-CT-3' \n",
    "    # 3'-CG-5' > 3'-GA-5' ==> 5'-GC-3' > 5'-AG-3'\n",
    "    if len(string) != 5 or string[2] != '>':\n",
    "        raise ValueError(\"Input string must be in the format 'NN>NN'\")\n",
    "    \n",
    "    string_pair = string.split('>')\n",
    "    new_string = reverse_complement(string_pair[0]) + '>' + reverse_complement(string_pair[1])\n",
    "    \n",
    "    return new_string\n",
    "\n",
    "dir=\"/mmfs1/gscratch/stergachislab/mhsohny/SMaHT/Improving_SomaticVariantCalling_through_DSA/Fiber-seq\"\n",
    "DSA=\"/mmfs1/gscratch/stergachislab/mhsohny/SMaHT/DSA/DSA_COLO829BL_v3.0.0.fasta\"\n",
    "\n",
    "DSA_pysam = pysam.FastaFile(f\"{DSA}\")\n",
    "\n",
    "colotb_snvs_pass_annot_FlaggerHap_glfilt_pileupfilt = read_vcf(f\"{dir}/VariantCalls_DeepVariant_1.6.1/Mutational_Spectrum/COLO829T_PassageB_DSA.deepvariant.split.snv.modified.final.vcf.gz\")\n",
    "colota_snvs_pass_annot_FlaggerHap_glfilt_pileupfilt = read_vcf(f\"{dir}/VariantCalls_DeepVariant_1.6.1/Mutational_Spectrum/COLO829T_PassageA_DSA.deepvariant.split.snv.modified.final.vcf.gz\")\n",
    "\n",
    "colotb_snvs_pass_annot_FlaggerHap_glfilt_pileupfilt['SNVid'] = colotb_snvs_pass_annot_FlaggerHap_glfilt_pileupfilt[['CHROM', 'POS', 'REF', 'ALT']].astype(str).apply('_'.join, axis=1)\n",
    "colota_snvs_pass_annot_FlaggerHap_glfilt_pileupfilt['SNVid'] = colota_snvs_pass_annot_FlaggerHap_glfilt_pileupfilt[['CHROM', 'POS', 'REF', 'ALT']].astype(str).apply('_'.join, axis=1)\n",
    "\n",
    "colotb_snvs_pass_annot_FlaggerHap_glfilt_pileupfilt_set = set(colotb_snvs_pass_annot_FlaggerHap_glfilt_pileupfilt['SNVid'].values)\n",
    "colota_snvs_pass_annot_FlaggerHap_glfilt_pileupfilt_set = set(colota_snvs_pass_annot_FlaggerHap_glfilt_pileupfilt['SNVid'].values)\n",
    "\n",
    "tba_FlaggerHap_glfilt_pileupfilt_tba_set = colotb_snvs_pass_annot_FlaggerHap_glfilt_pileupfilt_set.intersection(colota_snvs_pass_annot_FlaggerHap_glfilt_pileupfilt_set)\n",
    "tb_FlaggerHap_glfilt_pileupfilt_onlytb_set = colotb_snvs_pass_annot_FlaggerHap_glfilt_pileupfilt_set.difference(colota_snvs_pass_annot_FlaggerHap_glfilt_pileupfilt_set)\n",
    "ta_FlaggerHap_glfilt_pileupfilt_onlyta_set = colota_snvs_pass_annot_FlaggerHap_glfilt_pileupfilt_set.difference(colotb_snvs_pass_annot_FlaggerHap_glfilt_pileupfilt_set)\n",
    "\n",
    "print(f\"The number of Shared SNVs between Passage B and A: {len(tba_FlaggerHap_glfilt_pileupfilt_tba_set):,}\")\n",
    "print(f\"The number of B specific SNVs: {len(tb_FlaggerHap_glfilt_pileupfilt_onlytb_set):,}\")\n",
    "print(f\"The number of A specific SNVs: {len(ta_FlaggerHap_glfilt_pileupfilt_onlyta_set):,}\")\n",
    "\n",
    "print(f\"The number of DBS variants for Passage B: {colotb_snvs_pass_annot_FlaggerHap_glfilt_pileupfilt[colotb_snvs_pass_annot_FlaggerHap_glfilt_pileupfilt.groupby('CHROM')['POS'].diff() == 1].shape[0]:,}\")\n",
    "print(f\"The number of DBS variants for Passage B: {colotb_snvs_pass_annot_FlaggerHap_glfilt_pileupfilt[colotb_snvs_pass_annot_FlaggerHap_glfilt_pileupfilt.groupby('CHROM')['POS'].diff().shift(-1) == 1].shape[0]:,}\")\n",
    "\n",
    "print(f\"The number of DBS variants for Passage A: {colota_snvs_pass_annot_FlaggerHap_glfilt_pileupfilt[colota_snvs_pass_annot_FlaggerHap_glfilt_pileupfilt.groupby('CHROM')['POS'].diff() == 1].shape[0]:,}\")\n",
    "print(f\"The number of DBS variants for Passage A: {colota_snvs_pass_annot_FlaggerHap_glfilt_pileupfilt[colota_snvs_pass_annot_FlaggerHap_glfilt_pileupfilt.groupby('CHROM')['POS'].diff().shift(-1) == 1].shape[0]:,}\")\n",
    "\n",
    "print(f\"The number of DBS variants for Passage B: {colotb_snvs_pass_annot_FlaggerHap_glfilt_pileupfilt[colotb_snvs_pass_annot_FlaggerHap_glfilt_pileupfilt.groupby('CHROM')['POS'].diff() == 1].shape[0]:,}\")\n",
    "print(f\"The number of DBS variants for Passage B: {colotb_snvs_pass_annot_FlaggerHap_glfilt_pileupfilt[colotb_snvs_pass_annot_FlaggerHap_glfilt_pileupfilt.groupby('CHROM')['POS'].diff().shift(-1) == 1].shape[0]:,}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "def dbs78sig_maker(df: pd.core.frame.DataFrame, cram: pysam.libcalignmentfile.AlignmentFile, outdir: str, prefix: str) -> pd.core.frame.DataFrame:\n",
    "    \"\"\"\n",
    "    Parameters\n",
    "    ----------\n",
    "    \n",
    "    df : pandas.core.frame.DataFrame\n",
    "        vcf read through read_vcf()\n",
    "    cram : pysam.libcalignmentfile.AlignmentFile\n",
    "        cram file read using pysam\n",
    "    prefix : str\n",
    "        file name prefix\n",
    "    outdir : str\n",
    "        directory to write output\n",
    "    Write dbs78 input for SigProfiler\n",
    "\n",
    "    Example\n",
    "    ----------\n",
    "    MutationType    COLO829T_PassageB_DSA\n",
    "    AC>CA   1\n",
    "    AC>CG   0\n",
    "    AC>CT   0\n",
    "    \"\"\"\n",
    "\n",
    "    sampleid = df.columns[9]\n",
    "\n",
    "    dbs78_sigprofiler = ('AC>CA', 'AC>CG', 'AC>CT', 'AC>GA', 'AC>GG', 'AC>GT', 'AC>TA', 'AC>TG', 'AC>TT', 'AT>CA', 'AT>CC', 'AT>CG', 'AT>GA', 'AT>GC', 'AT>TA', 'CC>AA', 'CC>AG', 'CC>AT', 'CC>GA', 'CC>GG', 'CC>GT', 'CC>TA', 'CC>TG', 'CC>TT', 'CG>AT', 'CG>GC', 'CG>GT', 'CG>TA', 'CG>TC', 'CG>TT', 'CT>AA', 'CT>AC', 'CT>AG', 'CT>GA', 'CT>GC', 'CT>GG', 'CT>TA', 'CT>TC', 'CT>TG', 'GC>AA', 'GC>AG', 'GC>AT', 'GC>CA', 'GC>CG', 'GC>TA', 'TA>AT', 'TA>CG', 'TA>CT', 'TA>GC', 'TA>GG', 'TA>GT', 'TC>AA', 'TC>AG', 'TC>AT', 'TC>CA', 'TC>CG', 'TC>CT', 'TC>GA', 'TC>GG', 'TC>GT', 'TG>AA', 'TG>AC', 'TG>AT', 'TG>CA', 'TG>CC', 'TG>CT', 'TG>GA', 'TG>GC', 'TG>GT', 'TT>AA', 'TT>AC', 'TT>AG', 'TT>CA', 'TT>CC', 'TT>CG', 'TT>GA', 'TT>GC', 'TT>GG')\n",
    "    dbs78_sigprofiler = dict.fromkeys(dbs78_sigprofiler, 0)\n",
    "\n",
    "    df = df[df.groupby(\"CHROM\")['POS'].diff().shift(-1) == 1] # Preceding one of the two DBS SNVs\n",
    "\n",
    "    for index, row in tqdm(df.iterrows(), total=df.shape[0], desc=\"Processing Dinucleotides\"):\n",
    "        contig, start, end = row['CHROM'], row['POS'], row['POS']+1\n",
    "        \n",
    "        ref_dinucleotide = DSA_pysam.fetch(contig, start-1, end)\n",
    "\n",
    "        for read in cram.fetch(contig, start, end):\n",
    "            try:\n",
    "                ref_positions = read.get_reference_positions(full_length=True)\n",
    "                read_dinucleotide = read.query_sequence[ref_positions.index(start-1): ref_positions.index(end)]\n",
    "\n",
    "                if len(read_dinucleotide) == 2: # len(read_dinucleotide) != 2 -> insertion\n",
    "                    dbs = f\"{ref_dinucleotide}>{read_dinucleotide}\"\n",
    "                    \n",
    "                    if dbs in dbs78_sigprofiler:\n",
    "                        dbs78_sigprofiler[dbs] += 1\n",
    "                    elif dbs_context_change(dbs) in dbs78_sigprofiler:\n",
    "                        dbs78_sigprofiler[dbs_context_change(dbs)] += 1\n",
    "\n",
    "            except ValueError:\n",
    "                # -> deletion\n",
    "                pass\n",
    "    new_df = pd.DataFrame(dbs78_sigprofiler, index=[0]).T\n",
    "    new_df.columns = [sampleid]\n",
    "    new_df.index.name = \"MutationType\"\n",
    "    \n",
    "    new_df.to_csv(f\"{outdir}/{prefix}.DBS78.all\", sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing Dinucleotides: 100%|â–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆ| 1175/1175 [02:28<00:00,  7.90it/s]\n"
     ]
    }
   ],
   "source": [
    "# Passage B All\n",
    "colotb_cram = pysam.AlignmentFile(f\"{dir}/COLO829T_PassageB/COLO829T_PassageB_DSA.cram\", \"rc\")\n",
    "colotb_snvs_pass_annot_FlaggerHap_glfilt_pileupfilt\n",
    "\n",
    "dbs78sig_maker(colotb_snvs_pass_annot_FlaggerHap_glfilt_pileupfilt, colotb_cram, f\"{dir}/VariantCalls_DeepVariant_1.6.1/Mutational_Spectrum/02.DBS\", \"COLO829T_PassageB_DSA_All\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing Dinucleotides: 100%|â–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆ| 1097/1097 [01:48<00:00, 10.11it/s]\n"
     ]
    }
   ],
   "source": [
    "# Passage A All\n",
    "colota_cram = pysam.AlignmentFile(f\"{dir}/COLO829T_PassageA/COLO829T_PassageA_DSA.cram\", \"rc\")\n",
    "colota_snvs_pass_annot_FlaggerHap_glfilt_pileupfilt\n",
    "\n",
    "dbs78sig_maker(colota_snvs_pass_annot_FlaggerHap_glfilt_pileupfilt, colota_cram, f\"{dir}/VariantCalls_DeepVariant_1.6.1/Mutational_Spectrum/02.DBS\", \"COLO829T_PassageA_DSA_All\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sigPlt.plotDBS(matrix_path=f\"{dir}/VariantCalls_DeepVariant_1.6.1/Mutational_Spectrum/02.DBS/COLO829T_PassageB_DSA_All.DBS78.all\",\n",
    "               output_path=f\"{dir}/VariantCalls_DeepVariant_1.6.1/Mutational_Spectrum/02.DBS/All\",\n",
    "               project=\"COLO829T_PassageB_DSA_All.percentage\",\n",
    "               plot_type=\"78\",\n",
    "               savefig_format=\"pdf\",\n",
    "               percentage=True)\n",
    "\n",
    "sigPlt.plotDBS(matrix_path=f\"{dir}/VariantCalls_DeepVariant_1.6.1/Mutational_Spectrum/02.DBS/COLO829T_PassageB_DSA_All.DBS78.all\",\n",
    "               output_path=f\"{dir}/VariantCalls_DeepVariant_1.6.1/Mutational_Spectrum/02.DBS/All\",\n",
    "               project=\"COLO829T_PassageB_DSA_All\",\n",
    "               plot_type=\"78\",\n",
    "               savefig_format=\"pdf\",\n",
    "               percentage=False)\n",
    "\n",
    "sigPlt.plotDBS(matrix_path=f\"{dir}/VariantCalls_DeepVariant_1.6.1/Mutational_Spectrum/02.DBS/COLO829T_PassageA_DSA_All.DBS78.all\",\n",
    "               output_path=f\"{dir}/VariantCalls_DeepVariant_1.6.1/Mutational_Spectrum/02.DBS/All\",\n",
    "               project=\"COLO829T_PassageA_DSA_All.percentage\",\n",
    "               plot_type=\"78\",\n",
    "               savefig_format=\"pdf\",\n",
    "               percentage=True)\n",
    "\n",
    "sigPlt.plotDBS(matrix_path=f\"{dir}/VariantCalls_DeepVariant_1.6.1/Mutational_Spectrum/02.DBS/COLO829T_PassageA_DSA_All.DBS78.all\",\n",
    "               output_path=f\"{dir}/VariantCalls_DeepVariant_1.6.1/Mutational_Spectrum/02.DBS/All\",\n",
    "               project=\"COLO829T_PassageA_DSA_All\",\n",
    "               plot_type=\"78\",\n",
    "               savefig_format=\"pdf\",\n",
    "               percentage=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Assigning COSMIC sigs or Signature Database ...... \n",
      "|â–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆ| 1/1 [100%] in 0.0s (38.59/s) \n",
      "\n",
      "\n",
      " \n",
      "Your Job Is Successfully Completed! Thank You For Using SigProfilerAssignment.\n",
      " \n",
      "Assigning COSMIC sigs or Signature Database ...... \n",
      "|â–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆ| 1/1 [100%] in 0.0s (39.84/s) \n",
      "\n",
      "\n",
      " \n",
      "Your Job Is Successfully Completed! Thank You For Using SigProfilerAssignment.\n",
      " \n"
     ]
    }
   ],
   "source": [
    "os.system(f\"mkdir -p {dir}/VariantCalls_DeepVariant_1.6.1/Mutational_Spectrum/02.DBS/All/TB\") \n",
    "Analyze.cosmic_fit(\n",
    "    f\"{dir}/VariantCalls_DeepVariant_1.6.1/Mutational_Spectrum/02.DBS/COLO829T_PassageB_DSA_All.DBS78.all\", \n",
    "    f\"{dir}/VariantCalls_DeepVariant_1.6.1/Mutational_Spectrum/02.DBS/All/TB\", \n",
    "    input_type=\"matrix\", \n",
    "    context_type=\"DINUC\", \n",
    "    collapse_to_SBS96=False, \n",
    "    cosmic_version=3.4, \n",
    "    exome=False,\n",
    "    genome_build=\"GRCh38\", \n",
    "    signature_database=None,\n",
    "    exclude_signature_subgroups=None, \n",
    "    export_probabilities=True,\n",
    "    export_probabilities_per_mutation=False, \n",
    "    make_plots=True,\n",
    "    sample_reconstruction_plots=\"pdf\", \n",
    "    verbose=False)\n",
    "\n",
    "os.system(f\"mkdir -p {dir}/VariantCalls_DeepVariant_1.6.1/Mutational_Spectrum/02.DBS/All/TA\") \n",
    "Analyze.cosmic_fit(\n",
    "    f\"{dir}/VariantCalls_DeepVariant_1.6.1/Mutational_Spectrum/02.DBS/COLO829T_PassageA_DSA_All.DBS78.all\", \n",
    "    f\"{dir}/VariantCalls_DeepVariant_1.6.1/Mutational_Spectrum/02.DBS/All/TA\", \n",
    "    input_type=\"matrix\", \n",
    "    context_type=\"DINUC\", \n",
    "    collapse_to_SBS96=False, \n",
    "    cosmic_version=3.4, \n",
    "    exome=False,\n",
    "    genome_build=\"GRCh38\", \n",
    "    signature_database=None,\n",
    "    exclude_signature_subgroups=None, \n",
    "    export_probabilities=True,\n",
    "    export_probabilities_per_mutation=False, \n",
    "    make_plots=True,\n",
    "    sample_reconstruction_plots=\"pdf\", \n",
    "    verbose=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing Dinucleotides: 100%|â–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆ| 1013/1013 [02:19<00:00,  7.27it/s]\n",
      "Processing Dinucleotides: 100%|â–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆ| 1013/1013 [01:45<00:00,  9.64it/s]\n",
      "Processing Dinucleotides: 100%|â–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆ| 158/158 [00:13<00:00, 11.64it/s]\n",
      "Processing Dinucleotides: 100%|â–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆ| 75/75 [00:06<00:00, 11.17it/s]\n"
     ]
    }
   ],
   "source": [
    "sharedb = colotb_snvs_pass_annot_FlaggerHap_glfilt_pileupfilt[colotb_snvs_pass_annot_FlaggerHap_glfilt_pileupfilt['SNVid'].isin(tba_FlaggerHap_glfilt_pileupfilt_tba_set)]\n",
    "shareda = colota_snvs_pass_annot_FlaggerHap_glfilt_pileupfilt[colota_snvs_pass_annot_FlaggerHap_glfilt_pileupfilt['SNVid'].isin(tba_FlaggerHap_glfilt_pileupfilt_tba_set)]\n",
    "bspecific = colotb_snvs_pass_annot_FlaggerHap_glfilt_pileupfilt[colotb_snvs_pass_annot_FlaggerHap_glfilt_pileupfilt['SNVid'].isin(tb_FlaggerHap_glfilt_pileupfilt_onlytb_set)]\n",
    "aspecific = colota_snvs_pass_annot_FlaggerHap_glfilt_pileupfilt[colota_snvs_pass_annot_FlaggerHap_glfilt_pileupfilt['SNVid'].isin(ta_FlaggerHap_glfilt_pileupfilt_onlyta_set)]\n",
    "\n",
    "colotb_cram = pysam.AlignmentFile(f\"{dir}/COLO829T_PassageB/COLO829T_PassageB_DSA.cram\", \"rc\")\n",
    "colota_cram = pysam.AlignmentFile(f\"{dir}/COLO829T_PassageA/COLO829T_PassageA_DSA.cram\", \"rc\")\n",
    "\n",
    "dbs78sig_maker(sharedb, colotb_cram, f\"{dir}/VariantCalls_DeepVariant_1.6.1/Mutational_Spectrum/02.DBS\", \"COLO829T_PassageB_DSA_Shared\")\n",
    "dbs78sig_maker(shareda, colota_cram, f\"{dir}/VariantCalls_DeepVariant_1.6.1/Mutational_Spectrum/02.DBS\", \"COLO829T_PassageA_DSA_Shared\")\n",
    "\n",
    "dbs78sig_maker(bspecific, colotb_cram, f\"{dir}/VariantCalls_DeepVariant_1.6.1/Mutational_Spectrum/02.DBS\", \"COLO829T_PassageB_DSA_B-Specific\")\n",
    "dbs78sig_maker(aspecific, colota_cram, f\"{dir}/VariantCalls_DeepVariant_1.6.1/Mutational_Spectrum/02.DBS\", \"COLO829T_PassageA_DSA_A-Specific\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "sigPlt.plotDBS(matrix_path=f\"{dir}/VariantCalls_DeepVariant_1.6.1/Mutational_Spectrum/02.DBS/COLO829T_PassageB_DSA_Shared.DBS78.all\",\n",
    "               output_path=f\"{dir}/VariantCalls_DeepVariant_1.6.1/Mutational_Spectrum/02.DBS/Shared\",\n",
    "               project=\"COLO829T_PassageB_DSA_Shared.percentage\",\n",
    "               plot_type=\"78\",\n",
    "               savefig_format=\"pdf\",\n",
    "               percentage=True)\n",
    "\n",
    "sigPlt.plotDBS(matrix_path=f\"{dir}/VariantCalls_DeepVariant_1.6.1/Mutational_Spectrum/02.DBS/COLO829T_PassageB_DSA_Shared.DBS78.all\",\n",
    "               output_path=f\"{dir}/VariantCalls_DeepVariant_1.6.1/Mutational_Spectrum/02.DBS/Shared\",\n",
    "               project=\"COLO829T_PassageB_DSA_Shared\",\n",
    "               plot_type=\"78\",\n",
    "               savefig_format=\"pdf\",\n",
    "               percentage=False)\n",
    "\n",
    "sigPlt.plotDBS(matrix_path=f\"{dir}/VariantCalls_DeepVariant_1.6.1/Mutational_Spectrum/02.DBS/COLO829T_PassageB_DSA_B-Specific.DBS78.all\",\n",
    "               output_path=f\"{dir}/VariantCalls_DeepVariant_1.6.1/Mutational_Spectrum/02.DBS/B-Specific\",\n",
    "               project=\"COLO829T_PassageB_DSA_B-Specific.percentage\",\n",
    "               plot_type=\"78\",\n",
    "               savefig_format=\"pdf\",\n",
    "               percentage=True)\n",
    "\n",
    "sigPlt.plotDBS(matrix_path=f\"{dir}/VariantCalls_DeepVariant_1.6.1/Mutational_Spectrum/02.DBS/COLO829T_PassageB_DSA_B-Specific.DBS78.all\",\n",
    "               output_path=f\"{dir}/VariantCalls_DeepVariant_1.6.1/Mutational_Spectrum/02.DBS/B-Specific\",\n",
    "               project=\"COLO829T_PassageB_DSA_B-Specific\",\n",
    "               plot_type=\"78\",\n",
    "               savefig_format=\"pdf\",\n",
    "               percentage=False)\n",
    "\n",
    "sigPlt.plotDBS(matrix_path=f\"{dir}/VariantCalls_DeepVariant_1.6.1/Mutational_Spectrum/02.DBS/COLO829T_PassageA_DSA_Shared.DBS78.all\",\n",
    "               output_path=f\"{dir}/VariantCalls_DeepVariant_1.6.1/Mutational_Spectrum/02.DBS/Shared\",\n",
    "               project=\"COLO829T_PassageA_DSA_Shared.percentage\",\n",
    "               plot_type=\"78\",\n",
    "               savefig_format=\"pdf\",\n",
    "               percentage=True)\n",
    "\n",
    "sigPlt.plotDBS(matrix_path=f\"{dir}/VariantCalls_DeepVariant_1.6.1/Mutational_Spectrum/02.DBS/COLO829T_PassageA_DSA_Shared.DBS78.all\",\n",
    "               output_path=f\"{dir}/VariantCalls_DeepVariant_1.6.1/Mutational_Spectrum/02.DBS/Shared\",\n",
    "               project=\"COLO829T_PassageA_DSA_Shared\",\n",
    "               plot_type=\"78\",\n",
    "               savefig_format=\"pdf\",\n",
    "               percentage=False)\n",
    "\n",
    "sigPlt.plotDBS(matrix_path=f\"{dir}/VariantCalls_DeepVariant_1.6.1/Mutational_Spectrum/02.DBS/COLO829T_PassageA_DSA_A-Specific.DBS78.all\",\n",
    "               output_path=f\"{dir}/VariantCalls_DeepVariant_1.6.1/Mutational_Spectrum/02.DBS/A-Specific\",\n",
    "               project=\"COLO829T_PassageA_DSA_A-Specific.percentage\",\n",
    "               plot_type=\"78\",\n",
    "               savefig_format=\"pdf\",\n",
    "               percentage=True)\n",
    "\n",
    "sigPlt.plotDBS(matrix_path=f\"{dir}/VariantCalls_DeepVariant_1.6.1/Mutational_Spectrum/02.DBS/COLO829T_PassageA_DSA_A-Specific.DBS78.all\",\n",
    "               output_path=f\"{dir}/VariantCalls_DeepVariant_1.6.1/Mutational_Spectrum/02.DBS/A-Specific\",\n",
    "               project=\"COLO829T_PassageA_DSA_A-Specific\",\n",
    "               plot_type=\"78\",\n",
    "               savefig_format=\"pdf\",\n",
    "               percentage=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Assigning COSMIC sigs or Signature Database ...... \n",
      "|â–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆ| 1/1 [100%] in 0.2s (4.93/s) \n",
      "\n",
      "\n",
      " \n",
      "Your Job Is Successfully Completed! Thank You For Using SigProfilerAssignment.\n",
      " \n",
      "Assigning COSMIC sigs or Signature Database ...... \n",
      "|â–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆ| 1/1 [100%] in 0.0s (20.25/s) \n",
      "\n",
      "\n",
      " \n",
      "Your Job Is Successfully Completed! Thank You For Using SigProfilerAssignment.\n",
      " \n",
      "Assigning COSMIC sigs or Signature Database ...... \n",
      "|â–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆ| 1/1 [100%] in 0.0s (45.01/s) \n",
      "\n",
      "\n",
      " \n",
      "Your Job Is Successfully Completed! Thank You For Using SigProfilerAssignment.\n",
      " \n",
      "Assigning COSMIC sigs or Signature Database ...... \n",
      "|â–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆ| 1/1 [100%] in 0.1s (18.89/s) \n",
      "\n",
      "\n",
      " \n",
      "Your Job Is Successfully Completed! Thank You For Using SigProfilerAssignment.\n",
      " \n"
     ]
    }
   ],
   "source": [
    "# Shared (Passage B)\n",
    "os.system(f\"mkdir -p {dir}/VariantCalls_DeepVariant_1.6.1/Mutational_Spectrum/02.DBS/Shared/TB\") \n",
    "Analyze.cosmic_fit(\n",
    "    f\"{dir}/VariantCalls_DeepVariant_1.6.1/Mutational_Spectrum/02.DBS/COLO829T_PassageB_DSA_Shared.DBS78.all\", \n",
    "    f\"{dir}/VariantCalls_DeepVariant_1.6.1/Mutational_Spectrum/02.DBS/Shared/TB\", \n",
    "    input_type=\"matrix\", \n",
    "    context_type=\"DINUC\", \n",
    "    collapse_to_SBS96=False, \n",
    "    cosmic_version=3.4, \n",
    "    exome=False,\n",
    "    genome_build=\"GRCh38\", \n",
    "    signature_database=None,\n",
    "    exclude_signature_subgroups=None, \n",
    "    export_probabilities=True,\n",
    "    export_probabilities_per_mutation=False, \n",
    "    make_plots=True,\n",
    "    sample_reconstruction_plots=\"pdf\", \n",
    "    verbose=False)\n",
    "\n",
    "# B-Specific\n",
    "os.system(f\"mkdir -p {dir}/VariantCalls_DeepVariant_1.6.1/Mutational_Spectrum/02.DBS/B-Specific/TB\") \n",
    "Analyze.cosmic_fit(\n",
    "    f\"{dir}/VariantCalls_DeepVariant_1.6.1/Mutational_Spectrum/02.DBS/COLO829T_PassageB_DSA_B-Specific.DBS78.all\", \n",
    "    f\"{dir}/VariantCalls_DeepVariant_1.6.1/Mutational_Spectrum/02.DBS/B-Specific/TB\", \n",
    "    input_type=\"matrix\", \n",
    "    context_type=\"DINUC\", \n",
    "    collapse_to_SBS96=False, \n",
    "    cosmic_version=3.4, \n",
    "    exome=False,\n",
    "    genome_build=\"GRCh38\", \n",
    "    signature_database=None,\n",
    "    exclude_signature_subgroups=None, \n",
    "    export_probabilities=True,\n",
    "    export_probabilities_per_mutation=False, \n",
    "    make_plots=True,\n",
    "    sample_reconstruction_plots=\"pdf\", \n",
    "    verbose=False)\n",
    "\n",
    "# Shared (Passage A)\n",
    "os.system(f\"mkdir -p {dir}/VariantCalls_DeepVariant_1.6.1/Mutational_Spectrum/02.DBS/Shared/TA\") \n",
    "Analyze.cosmic_fit(\n",
    "    f\"{dir}/VariantCalls_DeepVariant_1.6.1/Mutational_Spectrum/02.DBS/COLO829T_PassageA_DSA_Shared.DBS78.all\", \n",
    "    f\"{dir}/VariantCalls_DeepVariant_1.6.1/Mutational_Spectrum/02.DBS/Shared/TA\", \n",
    "    input_type=\"matrix\", \n",
    "    context_type=\"DINUC\", \n",
    "    collapse_to_SBS96=False, \n",
    "    cosmic_version=3.4, \n",
    "    exome=False,\n",
    "    genome_build=\"GRCh38\", \n",
    "    signature_database=None,\n",
    "    exclude_signature_subgroups=None, \n",
    "    export_probabilities=True,\n",
    "    export_probabilities_per_mutation=False, \n",
    "    make_plots=True,\n",
    "    sample_reconstruction_plots=\"pdf\", \n",
    "    verbose=False)\n",
    "\n",
    "# A-Specific\n",
    "os.system(f\"mkdir -p {dir}/VariantCalls_DeepVariant_1.6.1/Mutational_Spectrum/02.DBS/A-Specific/TA\") \n",
    "Analyze.cosmic_fit(\n",
    "    f\"{dir}/VariantCalls_DeepVariant_1.6.1/Mutational_Spectrum/02.DBS/COLO829T_PassageA_DSA_A-Specific.DBS78.all\", \n",
    "    f\"{dir}/VariantCalls_DeepVariant_1.6.1/Mutational_Spectrum/02.DBS/A-Specific/TA\", \n",
    "    input_type=\"matrix\", \n",
    "    context_type=\"DINUC\", \n",
    "    collapse_to_SBS96=False, \n",
    "    cosmic_version=3.4, \n",
    "    exome=False,\n",
    "    genome_build=\"GRCh38\", \n",
    "    signature_database=None,\n",
    "    exclude_signature_subgroups=None, \n",
    "    export_probabilities=True,\n",
    "    export_probabilities_per_mutation=False, \n",
    "    make_plots=True,\n",
    "    sample_reconstruction_plots=\"pdf\", \n",
    "    verbose=False)\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "sigprofiler",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
